common/log/tag/interface.go

```diff
@@ -0,0 +1,192 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package tag
+
+import (
+	"time"
+)
+
+// Tag is the interface for logging system
+type Tag interface {
+	GetKey() string
+	GetValueType() valueTypeEnum
+
+	GetString() string
+	GetInteger() int64
+	GetDouble() float64
+	GetBool() bool
+	GetError() error
+	GetDuration() time.Duration
+	GetTime() time.Time
+	GetObject() interface{}
+}
+
+var _ Tag = (*tagImpl)(nil)
+
+// the tag value types supported
+const (
+	// ValueTypeString is string value type
+	ValueTypeString valueTypeEnum = 1
+	// ValueTypeInteger is any value of the types(int8,int16,int32,unit8,uint32,uint64) will be converted into int64
+	ValueTypeInteger valueTypeEnum = 2
+	// ValueTypeDouble is for both float and double will be converted into double
+	ValueTypeDouble valueTypeEnum = 3
+	// ValueTypeBool is bool value type
+	ValueTypeBool valueTypeEnum = 4
+	// ValueTypeError is error type value
+	ValueTypeError valueTypeEnum = 5
+	// ValueTypeDuration is duration type value
+	ValueTypeDuration valueTypeEnum = 6
+	// ValueTypeTime is time type value
+	ValueTypeTime valueTypeEnum = 7
+	// ValueTypeObject will be converted into string by fmt.Sprintf("%+v")
+	ValueTypeObject valueTypeEnum = 8
+)
+
+// keep this module private
+type valueTypeEnum int
+
+// keep this module private
+type tagImpl struct {
```

Comment:
+1 to Venkat's suggestion.

Reply:
That looks like a great idea.
So the reason behind this is we decided to solve the problem of tag conflict which has caused several ELIK ingestion issue in production which made us lost lots of logs. By using this interface, we want to forbidden people from using random string as key, and always generate tag from this logging module.

This idea should also work in the similar with. But only one concern: that it is still possible for people to get around of the restriction:
```
logger.Info("abc", tag.Tag(zap.String("randome name", "some value"))
```
I think it is still OK to make it weaker restriction but gain some perf benefits here. Also, I want to use
```
type Tag *zap.Field
```
With pointer it is even harder to hack into, and more efficient as zap.Field is a struct instead of interface. Without pointer we will end up creating a lot of small struct as well.
Also I will add some comments on this type to let people know that they should NOT do something like
```
field := zap.String("randome name", "some value")
logger.Info("abc", tag.Tag(&field)
```

Reply:
As Venkat mentioned, if you end up returning a pointer to a struct that will end up on the heap and have to be garbage collected. If you return a struct it all just remains on the stack and doesnt required garbage collection so is a lot more efficient for small structs. That is the reason why zap returns the struct rather than pointer in its corresponding implementation.. https://github.com/uber-go/zap/blob/master/field.go#L37

Reply:
@longquanzheng  - if you are concerned about callers overriding your restriction with tag.Tag(zap.String(...)), I would suggest the following instead of using *zap.Field as a type alias. 

Try
```
type struct Tag {
   field zap.Field
}
```

Because field is private, there is no way for callers outside your package to build Tag without calling one of our methods. 

As for "Without pointer we will end up creating a lot of small struct as well." 

It's ok to create these structs because these are returned by value instead of being returned as reference i.e. they will all be allocated in stack as opposed to heap, which is the optimization  that we are trying to achieve here. And that's also the reason Zap returns zap.Field by value and not by reference - https://github.com/uber-go/zap/blob/master/field.go#L100

Checkout this for more info - https://www.ardanlabs.com/blog/2017/05/language-mechanics-on-escape-analysis.html

Reply:
Got it, let's do this way.

Reply:
It think what you want to say is:
```
type Tag tagImpl

type tagImpl struct{
  field zap.Field
}
```